adapters/http-vertx-base/src/main/java/org/eclipse/hono/adapter/http/AbstractVertxBasedHttpProtocolAdapter.java

```diff
@@ -103,29 +103,45 @@ public final void start(Future<Void> startFuture) throws Exception {
         } else {
             eventSenderSupplier = (tenant, resultHandler) -> getHonoClient().getOrCreateEventSender(tenant, resultHandler);
             telemetrySenderSupplier = (tenant, resultHandler) -> getHonoClient().getOrCreateTelemetrySender(tenant, resultHandler);
+
+            Future<Void> preStartupFuture = Future.future();
+            preStartup(preStartupFuture);
+            Future<Void> httpServerStartupFuture = Future.future();
+            preStartupFuture.compose(ar -> {
+                bindHttpServer(createRouter(), httpServerStartupFuture);
+                connectToHono(null);
+            }, httpServerStartupFuture);
             
-            Future<Void> beforeHttpStartFuture = Future.future();
-            beforeHttpServerStart(beforeHttpStartFuture);
-            beforeHttpStartFuture.setHandler(ar -> {
-               if (ar.succeeded()) {
-                   bindHttpServer(createRouter(), startFuture);
-                   connectToHono(null);
-               } else {
-                   startFuture.fail(ar.cause());
-               }
+            httpServerStartupFuture.setHandler(ar -> {
+                if (ar.succeeded()) {
+                    postStartup();
+                    startFuture.complete();
+                } else {
+                    startFuture.fail(ar.cause());
+                }
             });
         }
     }
 
     /**
      * Invoked before the Http server is started.
      * May be overridden by sub-classes to provide additional startup handling.
+     * 
+     * @param startFuture  a future that has to be completed when this operation is finished
      */
-    protected void beforeHttpServerStart(Future<Void> startFuture) {
+    protected void preStartup(final Future<Void> startFuture) {
         startFuture.complete();
     }
 
     /**
+     * Invoked after startup of this Adapter is finished.
+     * May be overridden by sub-classes.
+     */
+    protected void postStartup() {
```

Comment:
I envisioned `postStartup()` to be invoked `after` the adapter startup has completed (as stated in the javadoc) (*), to provide a hook where e.g. startup logging might be done. In that sense the method isn't supposed to be able to let the adapter startup fail. Concerning exceptions, I would catch and log exceptions.

At the root of it, I still don't see much of a use for `postStartup()` for actual initialization logic, as it would mean having to deal with request handling logic possibly being invoked before `postStartup()` is finished. That is why I wouldn't wanna complicate things in `startup()` with handling yet another Future for that and rather keep things simple. We could also rename `postStartup()` to `onStartupSuccess()` to make it clearer.

---
(*) I see that the order of `postStartup()` and `startFuture.complete()` invocations should be switched for that still.